ospf6d/ospf6_auth_trailer.c

```diff
@@ -0,0 +1,842 @@
+	ospf6_auth->length = htons(auth_len);
+	ospf6_auth->reserved = 0;
+	ospf6_auth->id = htons(key_id);
+	ospf6_auth->seqnum_h = htonl(oi->at_cfg.seqnum_h);
+	ospf6_auth->seqnum_l = htonl(oi->at_cfg.seqnum_l);
+	memcpy(ospf6_auth->data, apad, hash_len);
+
+	ospf6_auth_update_digest(oi, oh, ospf6_auth, auth_str, auth_len,
+				 pkt_len, hash_algo);
+
+	if (CHECK_FLAG(oi->at_cfg.flags, OSPF6_AUTH_TRAILER_KEYCHAIN_VALID)) {
+		oi->at_cfg.hash_algo = KEYCHAIN_ALGO_NULL;
+		if (oi->at_cfg.auth_key) {
+			XFREE(MTYPE_OSPF6_AUTH_MANUAL_KEY, oi->at_cfg.auth_key);
+			oi->at_cfg.auth_key = NULL;
+		}
+
+		oi->at_cfg.key_id = 0;
+		UNSET_FLAG(oi->at_cfg.flags, OSPF6_AUTH_TRAILER_KEYCHAIN_VALID);
+	}
+}
+
+void ospf6_auth_update_digest(struct ospf6_interface *oi,
+			      struct ospf6_header *oh,
+			      struct ospf6_auth_hdr *ospf6_auth, char *auth_str,
+			      uint16_t auth_len, uint32_t pkt_len,
+			      enum keychain_hash_algo algo)
+{
+	uint32_t hash_len = keychain_get_hash_len(algo);
+	uint32_t block_s = keychain_get_block_size(algo);
+	uint32_t k_len = strlen(auth_str);
+	uint32_t ks_len = strlen(auth_str) + sizeof(CPID);
+	unsigned char ipad[block_s];
+	unsigned char opad[block_s];
+	unsigned char ko[block_s], ks[ks_len], tmp[hash_len];
+	unsigned char *first = NULL;
+	unsigned char *second = NULL;
+	unsigned char *first_mes, *second_mes;
+	unsigned char *first_hash, *second_hash;
+
+	memset(ko, 0, block_s);
+	memcpy(ks, auth_str, k_len);
+	memcpy(ks + k_len, &CPID, sizeof(CPID));
+	if (ks_len > hash_len) {
+		ospf6_hash_hmac_sha_digest(algo, ks, ks_len, tmp);
+		memcpy(ko, tmp, hash_len);
+	} else
+		memcpy(ko, ks, ks_len);
+
+	memcpy(ipad, ospf6_hash_ipad_max, block_s);
+	memcpy(opad, ospf6_hash_opad_max, block_s);
+
+	first = ospf6_hash_message_xor((unsigned char *)&ipad, ko, block_s);
+	second = ospf6_hash_message_xor((unsigned char *)&opad, ko, block_s);
+
+	first_mes = XMALLOC(MTYPE_OSPF6_AUTH_PKT, (block_s + pkt_len));
+	if (first_mes == NULL) {
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, first);
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, second);
+		zlog_err("Memory allocation for first message failed.");
+		return;
+	}
+
+	memcpy(first_mes, first, block_s);
+	memcpy(first_mes + block_s, oh, pkt_len);
+
+	first_hash = XMALLOC(MTYPE_OSPF6_AUTH_HASH, hash_len);
+	if (first_hash == NULL) {
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, first);
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, second);
+		XFREE(MTYPE_OSPF6_AUTH_PKT, first_mes);
+		zlog_err("Memory allocation for first hash failed.");
+		return;
+	}
+
+	ospf6_hash_hmac_sha_digest(algo, first_mes, (block_s + pkt_len),
+				   first_hash);
+
+	second_mes = XMALLOC(MTYPE_OSPF6_AUTH_PKT, (block_s + hash_len));
+	if (second_mes == NULL) {
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, first);
+		XFREE(MTYPE_OSPF6_AUTH_HASH_XOR, second);
+		XFREE(MTYPE_OSPF6_AUTH_PKT, first_mes);
+		XFREE(MTYPE_OSPF6_AUTH_HASH, first_hash);
+		zlog_err("Memory allocation for second message failed.");
```

Comment:
`XMALLOC` cannot fail by definition and will never return `NULL`

(affects all uses in multiple locations)

Reply:
Don't you think it's a best practice to always check for memory allocation failure. if system is running out of memory then malloc can fail. It's never guaranteed that malloc will succeed in all platforms right? I feel its good practise to check for malloc failure.

Reply:
@rameshabhinay perhaps you should spend a minute or two understanding the XMALLOC macros we have setup.  I would highly recommend looking at it.  As @eqvinox states it is impossible to have that function return NULL.  This code will never ever ever be executed.  Please go through and fix all instances of this in this PR.

Reply:
Let me go through the XMALLOC implementation once. 

@Donald Sharp
So no need to check for any memory allocation failure going further in future implementation?